File: libs/samsung-iot/samsung-iot.cpp
#include "pxt.h"
#include "IotService.h"

enum class SwitchState{
    off = 0,
    on = 1
};

enum class BulbTemp{
    soft_1 = 0,
    level_2 = 1,
    level_3 = 2,
    level_4 = 3,
    level_5 = 4,
    level_6 = 5,
    level_7 = 6,
    level_8 = 7,
    level_9 = 8,
    daylight_10 = 9
};

enum class BulbColour{
    white = 0,
    red = 1,
    orange = 2,
    yellow = 3,
    green = 4,
    blue = 5,
    purple = 6,
    violet = 7
};

/**
* share
* use
*/

IotService IotService(uBit.radio);

//% color=#000080 weight=100 icon="\uf0eb" block="Samsung IoT"
namespace samsungiot {

    int init() {
        return enableRadio();
    }

    int stringToNumber(ManagedString input)
    {
        ManagedString resultString = "";
        int result = 0;

        int l = input.length();
        for(int x=0; x < l; ++x)
        {
            if(input.charAt(x) != '\n' && input.charAt(x) != '\r')
            {
                if(isdigit(input.charAt(x)) || input.charAt(x) == '.')
                {
                    resultString = resultString + input.charAt(x);
                }
                else
                {
                    result = -1;
                    break;
                }
            }
        }
        if(result > -1)
        {
            float f = atof(resultString.toCharArray());
            result = round(f);
        }
        return result;
    }

    //%
    int setTempColour(BulbTemp colourTemp)
    {
        return (int)colourTemp;
    }

    //%
    int setLightColour(BulbColour lightcolour)
    {
        return (int)lightcolour;
    }

    //%
    void setBulbState(StringData* bulbName, SwitchState switchState)
    {
        init();

        ManagedString result = IotService.setBulbState(ManagedString(bulbName), (int)switchState);
        uBit.display.scroll(result);
    }

    //%
    bool getBulbState(StringData* bulbName)
    {
        bool state = true;
        init();

        ManagedString command = "bulbState/";
        command = command + bulbName;

        ManagedString result = IotService.getBulbState(command);
        uBit.display.scroll(result);
        if(strcmp(result.toCharArray(), "0") == 0)
        {
            state = false;
        }
        return state;
    }

    //%
    void setBulbLevel(StringData* bulbName, int level)
    {
        init();

        if(level < 0)
        {
            level = 0;
        }
        if(level > 100)
        {
            level = 100;
        }

        ManagedString result = IotService.setBulbLevel(ManagedString(bulbName), level);
        uBit.display.scroll(result);
    }

    //%
    int getBulbLevel(StringData* bulbName)
    {
        init();
        ManagedString command = "bulbLevel/";
        command = command + bulbName;

        ManagedString result = IotService.getBulbLevel(command);
        int level = stringToNumber(result);
        return level;
    }

    //%
    int getBulbTemp(StringData* bulbName)
    {
        init();
        ManagedString command = "bulbTemp/";
        command = command + bulbName;

        ManagedString result = IotService.getBulbTemp(command);
        int temp = stringToNumber(result);
        return temp;
    }

    //%
    void setBulbTemp(StringData* bulbName, BulbTemp temp)
    {
        init();
        if((int)temp < (int)BulbTemp::soft_1)
        {
            temp = BulbTemp::soft_1;
        }
        if((int)temp > (int)BulbTemp::daylight_10)
        {
            temp = BulbTemp::daylight_10;
        }

        ManagedString result = IotService.setBulbTemp(ManagedString(bulbName), (int)temp);
        uBit.display.scroll(result);
    }

    //%
    void setBulbNextTemp(StringData* bulbName)
    {
        int value = getBulbTemp(bulbName);
        ++value;
        if(value > (int)BulbTemp::daylight_10)
        {
            value = (int)BulbTemp::soft_1;
        }
        setBulbTemp(bulbName,(BulbTemp)value);
    }

    //%
    void setBulbColour(StringData* bulbName, BulbColour colour)
    {
        init();
        if((int)colour < (int)BulbColour::white)
        {
            colour = BulbColour::white;
        }
        if((int)colour > (int)BulbColour::violet)
        {
            colour = BulbColour::violet;
        }

        ManagedString result = IotService.setBulbColour(ManagedString(bulbName), (int)colour);
        uBit.display.scroll(result);
    }

    //%
    int getBulbColour(StringData* bulbName)
    {
        init();
        ManagedString command = "bulbColour/";
        command = command + bulbName;

        ManagedString result = IotService.getBulbColour(command);
        int colour = stringToNumber(result);
        return colour;
    }


    //%
    void setBulbNextColour(StringData* bulbName)
    {
        int value = getBulbColour(bulbName);
        ++value;
        if(value > (int)BulbColour::violet)
        {
            value = (int)BulbColour::white;
        }
        setBulbColour(bulbName,(BulbColour)value);
    }



    //%
    void turnSwitch(StringData* switchName, SwitchState switchState)
    {
        init();
        ManagedString result = IotService.setSwitchState(ManagedString(switchName),(int)switchState);
        uBit.display.scroll(result);
    }

    //%
    bool getSwitchState(StringData* switchName)
    {
        bool state = false;

        init();
        ManagedString command = "switchState/";
        command = command + switchName;

        ManagedString result = IotService.getSwitchState(command);
        if(strcmp(result.toCharArray(), "on") == 0)
        {
            state = true;
        }
        return state;
    }


    //%
    bool getSensorState(StringData* sensorName)
    {
        init();
        ManagedString command = "sensorState/";
        command = command + sensorName;
        ManagedString s = IotService.getSensorState(command);

        bool result = false;
        if(strcmp("active", s.toCharArray()) == 0)
        {
            result = true;
        }

        return result;
    }
};

File: libs/samsung-iot/samsung-iot.ts
//% color=#000080 weight=100 icon="\uf0eb" block="Samsung IoT"
namespace samsungiot {

    //% blockId=samsungiot_setTempColour block="colour temperature %colour" weight=66
    //% shim=samsungiot::setTempColour
    export function setTempColour(colourTemp:BulbTemp): number {
        return colourTemp;
    }

    //% blockId=samsungiot_setLightColour block="colour %colour" weight=66
    //% shim=samsungiot::setLightColour
    export function setLightColour(lightcolour:BulbColour): number {
        return lightcolour;
    }

    //% blockId=samsungiot_setBulbState block="turn bulb| %name |to %t" weight=60
    //% shim=samsungiot::setBulbState
    export function setBulbState(name:string, state: SwitchState): void{
        basic.showString("OK", 150);
        return;
    }

    //% blockId=samsungiot_getBulbState block="get bulb state %name" weight=64
    //% shim=samsungiot::getBulbState
    export function getBulbState(name:string): boolean {
        
        return Math.randomBoolean();
    }

    //% blockId=samsungiot_getBulbLevel block="get bulb level %name" weight=63
    //% shim=samsungiot::getBulbLevel
    export function getBulbLevel(name:string): number {
        return Math.random(101);
    }

    //% blockId=samsungiot_getBulbTemp block="get bulb colour temperature %name" weight=62
    //% shim=samsungiot::getBulbTemp
    export function getBulbTemp(name:string): number {
        return Math.random(10);
    }

    //% blockId=samsungiot_getBulbColour block="get bulb colour  %name" weight=62
    //% shim=samsungiot::getBulbColour
    export function getBulbColour(name:string): number {
        return Math.random(8);
    }

    //% blockId=samsungiot_setBulbLevel block="turn bulb| %name |to level %level" weight=59
    //% shim=samsungiot::setBulbLevel
    export function setBulbLevel(name:string, level: number): void{
        basic.showString("OK", 150);
        return;
    }

    //% blockId=samsungiot_setBulbTemp block="turn bulb| %name |to colour temperature %colour" weight=58
    //% shim=samsungiot::setBulbTemp
    export function setBulbTemp(name:string, colour: BulbTemp): void{
        basic.showString("OK", 150);
        return;
    }

    //% blockId=samsungiot_setBulbNextTemp block="turn bulb| %name |to next colour temperature" weight=58
    //% shim=samsungiot::setBulbNextTemp
    export function setBulbNextTemp(name:string): void{
        basic.showString("OK", 150);
        return;
    }

    //% blockId=samsungiot_setBulbColour block="turn bulb| %name |to colour  %colour" weight=58
    //% shim=samsungiot::setBulbColour
    export function setBulbColour(name:string, colour: BulbColour): void{
        basic.showString("OK", 150);
        return;
    }

    //% blockId=samsungiot_setBulbNextColour block="turn bulb| %name |to next colour" weight=58
    //% shim=samsungiot::setBulbNextColour
    export function setBulbNextColour(name:string): void{
        basic.showString("OK", 150);
        return;
    }

    //% blockId=samsungiot_turnSwitch block="turn switch| %name |to %t" weight=57
    //% shim=samsungiot::turnSwitch
    export function turnSwitch(name:string, state: SwitchState): void {
        basic.showString("OK", 150);
        return;
    }

    //% blockId=samsungiot_getSwitchState block="get switch state %name" weight=64
    //% shim=samsungiot::getSwitchState
    export function getSwitchState(name:string): boolean {
        return Math.randomBoolean();
    }


    //% blockId=samsungiot_getSensorState block="get sensor state| %name" weight=61
    //% shim=samsungiot::getSensorState
    export function getSensorState(name:string): boolean {
        return Math.randomBoolean();
    }

    //% blockId=samsungiot_getSensorTemp block="get sensor temperature| %name" weight=61
    //% shim=samsungiot::getSensorTemp
    export function getSensorTemp(name:string): number {
        return Math.random(25) + 10;
    }
}